src/System/Random/Internal.hs

```diff
@@ -0,0 +1,1040 @@
+instance UniformRange Double where
+  uniformRM (l, h) g = do
+    w64 <- uniformWord64 g
+    let x = word64ToDoubleInUnitInterval w64
+    return $ (h - l) * x + l
+
+-- | Turns a given uniformly distributed 'Word64' value into a uniformly
+-- distributed 'Double' value in the range [0, 1].
+word64ToDoubleInUnitInterval :: Word64 -> Double
+word64ToDoubleInUnitInterval w64 = d / m
+  where
+    d = fromIntegral w64 :: Double
+    m = fromIntegral (maxBound :: Word64) :: Double
+{-# INLINE word64ToDoubleInUnitInterval #-}
+
+-- | See [Floating point number caveats](System-Random-Stateful.html#fpcaveats).
+instance UniformRange Float where
+  uniformRM (l, h) g = do
+    w32 <- uniformWord32 g
+    let x = word32ToFloatInUnitInterval w32
+    return $ (h - l) * x + l
+
+-- | Turns a given uniformly distributed 'Word32' value into a uniformly
+-- distributed 'Float' value in the range [0,1].
+word32ToFloatInUnitInterval :: Word32 -> Float
+word32ToFloatInUnitInterval w32 = f / m
+  where
+    f = fromIntegral w32 :: Float
+    m = fromIntegral (maxBound :: Word32) :: Float
+{-# INLINE word32ToFloatInUnitInterval #-}
+
+-- The two integer functions below take an [inclusive,inclusive] range.
+randomIvalIntegral :: (RandomGen g, Integral a) => (a, a) -> g -> (a, g)
+randomIvalIntegral (l,h) = randomIvalInteger (toInteger l, toInteger h)
+
+{-# SPECIALIZE randomIvalInteger :: (Num a) =>
+    (Integer, Integer) -> StdGen -> (a, StdGen) #-}
+
+randomIvalInteger :: (RandomGen g, Num a) => (Integer, Integer) -> g -> (a, g)
+randomIvalInteger (l,h) rng
+ | l > h     = randomIvalInteger (h,l) rng
+ | otherwise = case f 1 0 rng of (v, rng') -> (fromInteger (l + v `mod` k), rng')
+     where
+       (genlo, genhi) = genRange rng
+       b = fromIntegral genhi - fromIntegral genlo + 1 :: Integer
+
+       -- Probabilities of the most likely and least likely result
+       -- will differ at most by a factor of (1 +- 1/q). Assuming the RandomGen
+       -- is uniform, of course
+
+       -- On average, log q / log b more pseudo-random values will be generated
+       -- than the minimum
+       q = 1000 :: Integer
+       k = h - l + 1
+       magtgt = k * q
+
+       -- generate pseudo-random values until we exceed the target magnitude
+       f mag v g | mag >= magtgt = (v, g)
+                 | otherwise = v' `seq`f (mag*b) v' g' where
+                        (x,g') = next g
+                        v' = v * b + (fromIntegral x - fromIntegral genlo)
+
+-- | Generate an integral in the range @[l, h]@ if @l <= h@ and @[h, l]@
+-- otherwise.
+uniformIntegralM :: (Bits a, Integral a, StatefulGen g m) => (a, a) -> g -> m a
+uniformIntegralM (l, h) gen = case l `compare` h of
+  LT -> do
+    let limit = h - l
+    let limitAsWord64 :: Word64 = fromIntegral limit
+    bounded <-
+      if fromIntegral limitAsWord64 == limit
+        -- Optimisation: if 'limit' fits into 'Word64', generate a bounded
```

Comment:
Use `Data.Bits.toIntegralSized`. It checks whether `a` is smaller than `Word64` first, and if it is, resolves the condition in compile time.

Reply:
The compile time optimisation won't bring any benefit here: this function is only used for `Natural` and `Integer`, so there are no compile time optimisations to be made.

Still, the function is neat, so I'll make a small change to use it.

Reply:
https://github.com/idontgetoutmuch/random/pull/167